pulser/waveforms.py

```diff
@@ -233,6 +269,39 @@ def __str__(self) -> str:
     def __repr__(self) -> str:
         return f'CompositeWaveform({self.duration} ns, {self._waveforms!r})'
 
+    def __getitem__(self, index_or_slice:
+                    Union[int, slice]) -> Union[float, np.ndarray]:
+        wf_start: int
+        if isinstance(index_or_slice, slice):
+            s: slice = self._check_slice(index_or_slice)
+            start: int = s.start
+            stop: int = s.stop
+            array: np.ndarray = np.array([])
+            wf_start = 0
+            for wf in self._waveforms:
+                if wf_start <= start and start <= wf_start+wf.duration:
+                    if wf_start <= stop and stop <= wf_start+wf.duration:
+                        return cast(np.ndarray,
+                                    wf.samples[start-wf_start:stop-wf_start])
+                    else:
+                        array = wf.samples[start-wf_start:]
+                if start < wf_start and stop > wf_start+wf.duration:
+                    array = np.concatenate((array, wf.samples))
+                if wf_start <= stop and stop <= wf_start+wf.duration:
+                    array = np.concatenate((array, wf.samples[:stop-wf_start]))
```

Comment:
I think it would be clearer to have the checks on `start` on the first stage and on `stop` on a second, like:
```suggestion
                elif start < wf_start:
                    if stop < wf_start:
                        break
                    elif stop <= wf_start+wf.duration:
                        array = np.concatenate((array, wf.samples[:stop-wf_start]))
                    else:
                        array = np.concatenate((array, wf.samples))
```
I think this even makes the if-clause above redundant.

Reply:
Note sure, here...

I built my code around three cases, that happen sequentially:
* The start index belongs to the current waveform
* The current waveform is completely included in the start-stop range
* The end index belongs to the current waveform

I'll add comments to make this appear more clearly.

Reply:
What I'm suggesting here is to take advantage of the previous check(s) on `start`, such that at this point (if using an if-elif construction) you know the start index can only belong to a previous waveform. As such, you only need to check:
1. If the stop index also belongs to a previous waveform, you just quit the loop.
2. If you got here, you know `stop > wf_start`, so you check if `stop < wf_start+wf.duration` (ie is within the current waveform) (and you can `break` here too btw)
3. If you got here, none of the above are true, so you know `stop > wf_start+wf.duration`

Note how the check 3. covers the case you are isolating above (`start < wf_start and stop > wf_start+wf.duration`)